sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala

```diff
@@ -59,11 +58,21 @@ abstract class CSVDataSource extends Serializable {
       sparkSession: SparkSession,
       inputPaths: Seq[FileStatus],
       parsedOptions: CSVOptions): Option[StructType]
+}
+
+object CSVDataSource {
+  def apply(options: CSVOptions): CSVDataSource = {
+    if (options.wholeFile) {
+      WholeFileCSVDataSource
+    } else {
+      TextInputCSVDataSource
+    }
+  }
 
   /**
    * Generates a header from the given row which is null-safe and duplicate-safe.
    */
-  protected def makeSafeHeader(
+  def makeSafeHeader(
```

Comment:
`makeSafeHeader` was moved from `CSVDataSource` class to `CSVDataSource` companion object so that this can be accessed in `DataFrameReader`.